[CSL-1542] Fix DNS lookups on Windows

adinapoli-iohk opened the pull request:
@rockbmb Let's start from here in debugging what's going on 😉 

Hopefully the CI will give us some insight!

adinapoli-iohk commented:
As predicted by @rockbmb the build on Win fails with:

```
C:\s\setup-exe-cache\x86_64-windows\Cabal-simple_Z6RU0evB_1.24.2.0_ghc-8.0.2.exe --builddir=.w\dist\ca59d0ab configure --with-ghc=C:\Users\appveyor\AppData\Local\Programs\stack\x86_64-windows\ghc-8.0.2\bin\ghc.EXE --with-ghc-pkg=C:\Users\appveyor\AppData\Local\Programs\stack\x86_64-windows\ghc-8.0.2\bin\ghc-pkg.EXE --user --package-db=clear --package-db=global --package-db=C:\s\snapshots\6dde1a60\pkgdb --package-db=C:\w\.w\install\5600629f\pkgdb --libdir=C:\w\.w\install\5600629f\lib --bindir=C:\w\.w\install\5600629f\bin --datadir=C:\w\.w\install\5600629f\share --libexecdir=C:\w\.w\install\5600629f\libexec --sysconfdir=C:\w\.w\install\5600629f\etc --docdir=C:\w\.w\install\5600629f\doc\cardano-sl-infra-0.5.1 --htmldir=C:\w\.w\install\5600629f\doc\cardano-sl-infra-0.5.1 --haddockdir=C:\w\.w\install\5600629f\doc\cardano-sl-infra-0.5.1 --dependency=QuickCheck=QuickCheck-2.9.2-LvlomhGMxrFBkX1zwiWpQu --dependency=aeson=aeson-1.1.2.0-bvg0hqcciCGZuD2wrCkaE --dependency=base=base-4.9.1.0 --dependency=base64-bytestring=base64-bytestring-1.0.0.1-4O9TExPPz6eI82Yj7L4SIj --dependency=binary=binary-0.8.3.0 --dependency=bytestring=bytestring-0.10.8.1 --dependency=cardano-report-server=cardano-report-server-0.2.1-8vzNOD1flKgISdU7x5q1rz --dependency=cardano-sl-core=cardano-sl-core-0.5.1-KShiPpIE76jJSCFjHYe3yN --dependency=cardano-sl-db=cardano-sl-db-0.5.1-bff1ALxj923KMbka9WuHv --dependency=containers=containers-0.5.7.1 --dependency=data-default=data-default-0.7.1.1-Atv1ZD6t1rI1MrH9qSGaQC --dependency=directory=directory-1.3.1.0-GjCsKrj9z3QCjDkUZfYR6V --dependency=dns=dns-2.0.12-EXkcbyPkz3tHIwXv0AbpKN --dependency=either=either-4.4.1.1-6PiwKYkn4v6B4KO2R2Fu1b --dependency=ekg-core=ekg-core-0.1.1.2-G1KRQAgpWXyCIS5PiHAudz --dependency=ether=ether-0.5.1.0-DQr71rwEgvNAXgHaobJvfV --dependency=exceptions=exceptions-0.8.3-ADEDRATtUrMK1JW4zMQ9U --dependency=filepath=filepath-1.4.1.1 --dependency=formatting=formatting-6.2.4-99Nwhnkkgy8Ff6qEBpTZer --dependency=generic-arbitrary=generic-arbitrary-0.1.0-4wdGlh95V1850BmN1AoxEe --dependency=hashable=hashable-1.2.6.1-2ZLNuHq395GGIHwEHuqZol --dependency=http-client=http-client-0.5.7.0-8ZNq6YJy5WC5GRhAya7J2c --dependency=http-client-tls=http-client-tls-0.3.5.1-FqbkarcvLviKyPVSs9N8PK --dependency=iproute=iproute-1.7.1-9BRunn4fZ2LJijFUX0wAoi --dependency=kademlia=kademlia-1.1.0.1-6w95LQQ6WrUEXRVRX5h4L3 --dependency=lens=lens-4.15.4-3P7PSyaTsWRLQnkSsAPtwp --dependency=list-t=list-t-1.0.0.1-5hQv0wfkuYf598er1kAI66 --dependency=log-warper=log-warper-1.1.2-Clrlrl9lV8zGLj0U6HjaHt --dependency=mmorph=mmorph-1.0.9-5bASEvmIkr5CpdAYXwPsn0 --dependency=monad-control=monad-control-1.0.2.2-1Ci8dYJ8Tb7E3HTcTqOb8l --dependency=mtl=mtl-2.2.1-BLKBelFsPB3BoFeSWSOYj6 --dependency=network-info=network-info-0.2.0.8-3CNHbSDxkB04OXdr5Q0x5c --dependency=network-transport=network-transport-0.5.2-C5qvioPDY0j34GsMszZ4XS --dependency=network-transport-tcp=network-transport-tcp-0.5.1-FlqVomLmHW7CB3XyTH9Qnm --dependency=node-sketch=node-sketch-0.2.0.0-8Xxhb4kDyXLDZM617eMDiX --dependency=optparse-applicative=optparse-applicative-0.13.2.0-bmydd61wXl1i03BWH5vZ --dependency=parsec=parsec-3.1.11-1aqiimsgq2dCv7eoaC8WuX --dependency=reflection=reflection-2.1.2-Bem12O1xFGSHr4C2Hyu1c4 --dependency=serokell-util=serokell-util-0.4.0-II7w27nVBoCHVZshBAYV1 --dependency=stm=stm-2.4.4.1-JQn4hNPyYjP5m9AcbI88Ve --dependency=tagged=tagged-0.8.5-1mTloBSoUxv8dqUr8XBGBt --dependency=template-haskell=template-haskell-2.11.1.0 --dependency=text=text-1.2.2.2-9UQZjEJZQFSGMffj1Z5g00 --dependency=text-format=text-format-0.3.1.1-3cyN5zsAtZ9A1Nt0xXDPiM --dependency=time=time-1.6.0.1 --dependency=time-units=time-units-1.0.0-LG1KGQdbetIIh4DlrSxGpy --dependency=transformers=transformers-0.5.2.0 --dependency=transformers-base=transformers-base-0.4.4-GKSs0LKukAL8ymq2naFmrH --dependency=transformers-lift=transformers-lift-0.2.0.1-CVMYJkCDI9QHJ8IjqDaT9i --dependency=universum=universum-0.4.3-8wlDcsohrXPLnujkMONGcY --dependency=unordered-containers=unordered-containers-0.2.8.0-Bp9XgxjuHxcI4tFehVMDGC --dependency=yaml=yaml-0.8.23.3-6GQCSkhKSH0HFzuN6hAQD7 --ghc-options -Werror --ghc-options -DCONFIG=qanet_tns_wallet --ghc-options -DGITREV=76b30c8fe97b57e81197df25a2f55a5d9b62022b --extra-include-dirs=C:\OpenSSL-Win64-v102\include --extra-include-dirs=C:\Users\appveyor\AppData\Local\Programs\stack\x86_64-windows\msys2-20150512\mingw64\include --extra-include-dirs=c:\w\rocksdb\include --extra-lib-dirs=C:\OpenSSL-Win64-v102 --extra-lib-dirs=C:\Users\appveyor\AppData\Local\Programs\stack\x86_64-windows\msys2-20150512\mingw64\bin --extra-lib-dirs=C:\Users\appveyor\AppData\Local\Programs\stack\x86_64-windows\msys2-20150512\mingw64\lib --extra-lib-dirs=c:\w --enable-tests --enable-benchmarks
    Process exited with code: ExitFailure 1
    Logs have been written to: C:\w\.w\logs\cardano-sl-infra-0.5.1.log
    Configuring cardano-sl-infra-0.5.1...
    Cabal-simple_Z6RU0evB_1.24.2.0_ghc-8.0.2.exe: Missing dependency on a foreign
    library:
    * Missing C library: iphlpapi
    This problem can usually be solved by installing the system package that
    provides this library (you may need the "-dev" version). If the library is
    already installed but in a non-standard location then you can use the flags
    --extra-include-dirs= and --extra-lib-dirs= to specify where it is.
Sorry, we tried running command for 5 times and all attempts were unsuccessful!
```

Next step would be trying to isolate the differences between the [very simple](https://github.com/input-output-hk/dns-resolver-windows/blob/master/appveyor.yml) appveyor.yaml we have committed in the standalone example vs the one [currently in use](https://github.com/input-output-hk/cardano-sl/blob/master/appveyor.yml) to see if anything obvious come up.

adinapoli-iohk commented:
Oh dear, this is weird -- I was fiddling with Appveyor in my standalone app (to test the platform thing) and now the tests fails randomly with a timeout error, even if I haven't changed the code! 😱 

https://ci.appveyor.com/project/adinapoli-iohk/dns-resolver-windows/history

rockbmb commented:
@adinapoli-iohk Can you tell me what's going on? DId the code I wrote for CSL-1481 stop working all of a sudden?

adinapoli-iohk commented:
@rockbmb Absolutely (sorry for keeping you out of the loop in the first place).

In a nutshell, @fersel reported the DNS code was not working on Windows and @dcoutts asked to find out more. Unfortunately I don't have the exact details about the breakage (aka I lack reproduction steps, or a windows machine to begin with). Duncan suggested to create a standalone example which would demonstrate the DNS code was working in a "simplified" environment, and so I did:

https://github.com/input-output-hk/dns-resolver-windows

Such app comes with 2 tests, one testing specifically your C code, the other testing a multi-value lookup. Both _used_ to work just fine but today I'm having random issues with Appveyor and the multi-lookup value is failing with a `TimeoutException` (but the C code is working just fine).

What I am trying now to understand is how so we can build things correctly in the standalone example (i.e. passing the `extra-library`) whereas for `cardano-sl-infra` this is not the case.

I hope this clarifies the state of affairs a little bit better, sorry!

rockbmb commented:
@adinapoli-iohk Thanks!

In Slack you mentioned the commit for `cardano-sl' for which I tested the DNS lookup, there is none (should've left it there..). I tested it locally by adding several calls to `getDefaultWindowsDnsServer` inside `/node/src/node/Main.hs`'s `main` function and started up the node while checking the console for IP addresses. It worked.

rockbmb commented:
@adinapoli-iohk Also, I can tell you straight away - the standalone repo won't work on Windows unless you `copy C:\Windows\System32\IPHLPAPI.dll _project's_directory_`

adinapoli-iohk commented:
>  Also, I can tell you straight away - the standalone repo won't work on Windows unless you copy C:\Windows\System32\IPHLPAPI.dll _project's_directory_

But this is the crazy part! It _used_ to work! Look here:

https://ci.appveyor.com/project/adinapoli-iohk/dns-resolver-windows/build/1.0.14

rockbmb commented:
@adinapoli-iohk that's a good idea, although when I was using `procmon` to check where `cabal` was looking for the libraries, it searched in `C:\Windows\System32` without the `--extra-lib-dirs` flag.

Also, there's something else I noticed - the tests fail. For them to fail, they must be run; for them to run, they must have been compiled; for them to have been compiled, that means the whole project was built. As such, there was no issue with finding `iphlpapi`. Perhaps the cause of failure is completely unrelated to CSL-1542

adinapoli-iohk commented:
> As such, there was no issue with finding iphlpapi

Yes exactly, this is why I'm puzzled to observe this difference in behaviour between the two environments (simplified vs cardano)

> that's a good idea, although when I was using procmon to check where cabal was looking for the libraries, it searched in C:\Windows\System32 without the --extra-lib-dirs flag.

Interesting. I guess it won't make any harm adding such explicit `--extra-lib-dirs` (although this won't explain _why_ is not necessary for the standalone app). More invasive solutions would be to extend the appveyor script to print out all the DLL registered in the system or the content of `System32`, to be sure such DLL is really there to begin with.

rockbmb commented:
@adinapoli-iohk Hm, it's quite unfortunate that CI isn't as deterministic as we'd like.

This failure in the latest commit of the standalone project is most likely a problem with AppVeyor itself, unless changing `google.it` to `google.co.uk` had some adverse effect.

What do you think of reverting that change just to make sure? If it still doesn't work, something's wrong with i.e. AppVeyor's cache.

adinapoli-iohk commented:
> This failure in the latest commit of the standalone project is most likely a problem with AppVeyor itself, unless changing google.it to google.co.uk had some adverse effect.

Totally agree -- I was getting desperate and I thought about some twisted plot whether traffic to that DNS would get blocked by Appveyor or stuff like that, but it's unrealistic. Let me try to revert the changes first, and wipe the cache (which I think I've never done before AND we are not even using it, now that I think about it) and see if I can observe any changes.

rockbmb commented:
@adinapoli-iohk I added a small AppVeyor-only test to CI (and removed the `extra-library: iphlpapi` section temporarily to make it work), and it seems to work. Maybe we can build, on top of this module, progressively more elaborate tests to try and weed out the source of the error.

What do you think?

adinapoli-iohk commented:
> Maybe we can build, on top of this module, progressively more elaborate tests to try and weed out the source of the error. What do you think?

That's a very good start, although I'm still flabbergasted that not only isn't the `extra-library` required, but that adding it throws an error. We'll try to shed some light on that later, if needed.

I think what you have there is solid, but to be sure to be in a strong position this isn't quite enough. We need to include tests for multi-value lookup, as @fersel is reporting the following:

> Everything's working with raw ip addresses, e.g. if my topology looks like the following:

```wallet:
  relays: [[{"addr": "52.59.160.100", "port": 3000}, {"addr": "52.58.254.151", "port": 3000}]]
```
> But if I change it to the dns corresponding to those ip addresses (also `addr` is changed to `host` as stated in the topology doc):

```wallet:
  relays: [[{"host": "cardano-node-0.aws.iohkdev.io", "port": 3000}, {"host": "cardano-node-1.aws.iohkdev.io", "port": 3000}]]
```
> node doesn't find relays anymore (I don't recall the exact problem, will check it a bit later)

This suggests indeed that DNSes are not resolving correctly. Mind, this might not have anything to do with the code you wrote, but we a bug in the `dns` library itself (which doesn't seem to be thoroughly tested on Windows), so we might need to put our Sherlock Holmes' hat and dig deeper 😉 

Let me add some more tests and let's see what the CI has to say about them!

adinapoli-iohk commented:
I've managed to nag Mr @fersel to run the standalone app (which I have enriched with debug info), and this is the report:

```
*** DEBUG INFO ***

DEFAULT DNS: Just "192.168.1.1"
www.google.it resolved to: Right [173.194.113.175,173.194.113.183,173.194.113.191,173.194.113.184]
cardano-node-0.aws.iohkdev.io resolved to: Left TimeoutExpired
cardano-node-1.aws.iohkdev.io resolved to: Left TimeoutExpired

*** END OF DEBUG INFO ***

Tests
  Unit tests
    getDefaultDnsServer works: OK
    multi-value lookups work:  FAIL
      Exception: user error (TimeoutExpired)
    resolving IOHK DNS works:  FAIL
      Exception: user error (TimeoutExpired)

2 out of 3 tests failed (0.36s)

Completed 2 action(s).
Test suite failure for package dns-resolver-windows-0.1.0.0
    dns-resolver-windows-test:  exited with: ExitFailure 1
Logs printed to console
```

So, report from the trenches:

* The code @rockbmb wrote definitely works; we have seen in his tests plus the fact the standalone app reports `192.168.1.1`.

* Failures seems to be non-deterministic: You can see how the first call to `lookup` for "www.google.it" works but all the other do not. In case you are wondering, here is how the tests are defined in the standalone examples:

```
main :: IO ()
main = do
  seed <- newResolvConf >>= makeResolvSeed
  -- Print some debug info
  putStrLn $ "\n*** DEBUG INFO ***\n"
  defaultDNS <- getDefaultDnsServer
  putStrLn $ "DEFAULT DNS: " <> show defaultDNS
  googIP <- withResolver seed $ \resolver -> DNS.lookup resolver "www.google.it" A
  c1     <- withResolver seed $ \resolver -> DNS.lookup resolver "cardano-node-0.aws.iohkdev.io" A
  c2     <- withResolver seed $ \resolver -> DNS.lookup resolver "cardano-node-1.aws.iohkdev.io" A
  putStrLn $ "www.google.it resolved to: " <> show googIP
  putStrLn $ "cardano-node-0.aws.iohkdev.io resolved to: " <> show c1
  putStrLn $ "cardano-node-1.aws.iohkdev.io resolved to: " <> show c2

  putStrLn $ "\n*** END OF DEBUG INFO ***\n"
  -- Run the tests
  defaultMain tests

tests :: TestTree
tests = testGroup "Tests" [unitTests]

unitTests :: TestTree
unitTests = testGroup "Unit tests"
  [ testCase "getDefaultDnsServer works" testGetDefaultDnsServer
  , testCase "multi-value lookups work"  testLookups
  , testCase "resolving IOHK DNS works"  testLookupsIOHK
  ]
[...] -- Rest
```

So we effectively try to resolve "www.google.it" twice: once in the debug info section, once in the tests. Why we are getting failures? In case you are wondering, I'm using very generous timeouts in the standalone app:

https://github.com/input-output-hk/dns-resolver-windows/blob/master/src/Network/DNS/Domains.hs#L51-L58

adinapoli-iohk commented:
@rockbmb Bingo!

![screen shot 2017-08-30 at 11 25 29](https://user-images.githubusercontent.com/29383371/29865505-fccd0e9a-8d75-11e7-921c-d2217f2a42ba.png)

adinapoli-iohk commented:
I have forked the `dns` package and I have started hacking on it to improve the Windows situation. I was successfully able to replicate failures:

https://ci.appveyor.com/project/adinapoli-iohk/dns/build/1.0.2

You will see the error is now different and that's because [I have added a more-specific exception](https://github.com/input-output-hk/dns/commit/bd899f6239cffef46cdb1b7509e733127720da73).

adinapoli-iohk commented:
Ok! I think I have fixed it! See here:

https://ci.appveyor.com/project/adinapoli-iohk/dns/build/1.0.4

Apart from the doctest failures, you will notice now that the DNS lookup works correctly (at least in this iteration of the build). The culprit seemed to be the custom version of `sendAll`, which I have now removed. I'm caution to declare victory yet, but this looks promising..

adinapoli-iohk commented:
Yay!

https://ci.appveyor.com/project/adinapoli-iohk/dns-resolver-windows/branch/master

dcoutts commented on node/test/Test/Pos/Network/Windows/DnsDomainsSpec.hs:
```diff
@@ -14,23 +15,35 @@ module Test.Pos.Network.Windows.DnsDomainsSpec
+    describe "Multi-value lookups" $ do
+        it "www.google.it resolves correctly" $ testLookupFor "www.google.it"
+        it "cardano-node-0.aws.iohkdev.io resolves correctly" $ testLookupFor "cardano-node-0.aws.iohkdev.io"
+        it "cardano-node-1.aws.iohkdev.io resolves correctly" $ testLookupFor "cardano-node-1.aws.iohkdev.io"
```
This may not be a robust test since it depends on our devops DNS setup.

Also, this does not really test multi-valued lookup.

dcoutts commented on stack.yaml:
```diff
@@ -81,6 +81,12 @@ packages:
+# Forked version of "dns" with proper Windows support
+- location:
+    git: https://github.com/input-output-hk/dns.git
+    commit: 6da3cd051bd6f6415c8ec3b53f1c0372a80a4376
```
Ok, but lets also try and get a minimal set of DNS package changes, suitable to go upstream.

This is OK for now, but next step should be to prepare a clean upstream PR.

adinapoli-iohk replied:
Aye, I'm on it as we speak (to be precise, I'm waiting for the tests to fully pass to be even surer my fix is hitting the nail on the head).

rockbmb commented:
@adinapoli-iohk I wasn't here when you fixed this, but excellent work! I wouldn't have found this error by myself.
